src/gui_common/TweakedColourPicker.cs

```diff
@@ -0,0 +1,345 @@
+        presets.Add(preset);
+        presetsContainer.AddChild(preset);
+
+        // Add preset to base class
+        base.AddPreset(color);
+    }
+
+    public new void ErasePreset(Color color)
+    {
+        OnPresetDeleted(presets.First(p => p.Color == color));
+    }
+
+    private void UpdateTooltips()
+    {
+        pickerButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_PICK_COLOR");
+        addPresetButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_ADD_PRESET");
+        hsvCheckButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_HSV_BUTTON_TOOLTIP");
+        rawCheckButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_RAW_BUTTON_TOOLTIP");
+
+        if (HsvMode)
+        {
+            sliderROrH.HintTooltip = TranslationServer.Translate("COLOR_PICKER_H_TOOLTIP");
+            sliderGOrS.HintTooltip = TranslationServer.Translate("COLOR_PICKER_S_TOOLTIP");
+            sliderBOrV.HintTooltip = TranslationServer.Translate("COLOR_PICKER_V_TOOLTIP");
+        }
+        else
+        {
+            sliderROrH.HintTooltip = TranslationServer.Translate("COLOR_PICKER_R_TOOLTIP");
+            sliderGOrS.HintTooltip = TranslationServer.Translate("COLOR_PICKER_G_TOOLTIP");
+            sliderBOrV.HintTooltip = TranslationServer.Translate("COLOR_PICKER_B_TOOLTIP");
+        }
+
+        sliderA.HintTooltip = TranslationServer.Translate("COLOR_PICKER_A_TOOLTIP");
+    }
+
+    private void UpdateButtonsState()
+    {
+        if (hsvCheckButton == null)
+            return;
+
+        hsvCheckButton.Pressed = HsvMode;
+        rawCheckButton.Pressed = RawMode;
+
+        hsvCheckButton.Disabled = !hsvButtonEnabled || RawMode;
+        rawCheckButton.Disabled = !rawButtonEnabled || HsvMode;
+    }
+
+    private void OnAddPresetButtonPressed()
+    {
+        AddPreset(Color);
+    }
+
+    private void OnPresetSelected(Color color)
+    {
+        Color = color;
+        EmitSignal("color_changed", Color);
+    }
+
+    private void OnPresetDeleted(TweakedColorPickerPreset preset)
+    {
+        presets.Remove(preset);
+        presetsContainer.RemoveChild(preset);
+        base.ErasePreset(preset.Color);
+        preset.QueueFree();
+    }
+
+    private void OnHSVButtonToggled(bool isOn)
+    {
+        HsvMode = isOn;
+    }
+
+    private void OnRawButtonToggled(bool isOn)
+    {
+        RawMode = isOn;
+    }
+
+    private void OnColorChanged(Color color)
+    {
+        hexColorEdit.Text = color.a8 == 255 ? color.ToHtml(false) : color.ToHtml(true);
+    }
+
+    /// <summary>
+    ///   Called when (keyboard) entered in hexColorEdit.
+    ///   Set Color when text is valid; reset if not.
+    ///   Because text_entered signal requires a string parameter but focus_exited zero, wrote lite this.
+    /// </summary>
+    /// <param name="color">Current hexColorEditor text</param>
+    private void OnHexColorChanged(string color = null)
```

Comment:
I find that if you click at a container, the LineEdit doesn't lose focus.

Reply:
It might be due to that there are some extra steps needed to fully defocus the line edit. I looked up Godot's color picker implementation and there's an extra `select(0, 0)` call that I assume is used to completely deselect the line edit on focus exit. Maybe this can be tried out to see if the line edit will behave correctly.

```c++
void ColorPicker::_focus_exit() {
	for (int i = 0; i < 4; i++) {
		if (!values[i]->get_line_edit()->get_menu()->is_visible())
			values[i]->get_line_edit()->select(0, 0);
	}
	c_text->select(0, 0);
}

void ColorPicker::_html_focus_exit() {
	if (c_text->get_menu()->is_visible())
		return;
	_html_entered(c_text->get_text());
	_focus_exit();
}
```
Not sure why they specifically chose [select](https://docs.godotengine.org/en/stable/classes/class_lineedit.html#class-lineedit-method-select) method instead of [deselect](https://docs.godotengine.org/en/stable/classes/class_lineedit.html#class-lineedit-method-deselect) as it seem to be a perfect fit there though. But I think both works exactly the same and for our use `deselect` should be used for its clarity.

Reply:
Also if this works, you may want to connect the text_entered and focus_exited signal to separate callbacks as otherwise it would probably be a bit annoying if the line edit completely deselects every time you press enter.

Reply:
Well I think that when you click at a container or something it **can't get focus**, so the color picker **can't** lose focus. That's the problem. I'll try using _Input().

Reply:
I don't entirely understand what the problem here is, so the LineEdit doesn't lose focus when you click Container controls? To my knowledge that's totally expected as those aren't even interactable and they don't grab focus now that reread what you said. And when LineEdit does lose focus to other control, it deselects by itself.

I think the new _Input approach is even more unnecessarily more lines of code now than the one I suggested and I don't see any change with the focus behavior from testing?

Reply:
I know that it is easier to just check if the text is valid when exiting, but I do think that it is crazy if you can even key in Chinese or other characters in a hex-only line edit. So anyway I choose to filter the input.
![image](https://user-images.githubusercontent.com/60734649/130349016-34cc04d8-0525-443e-8c3e-497905090ed7.png)

Reply:
I even think that it is better to just make it read-only. The HTML color edit doesn't worth that much attention. It's too professional for most players.

Reply:
It can still be useful for some people.

> I know that it is easier to just check if the text is valid when exiting, but I do think that it is crazy if you can even key in Chinese or other characters in a hex-only line edit. So anyway I choose to filter the input.

I was suggesting a more simpler solution but if you think this is a better way then I guess we could try going with this. Though it should be at least as robust as possible and handles edge cases pretty well, as for example I tried testing a bit and looks like there's a bug where if you tried to copy paste a string with non-hex characters to the line edit, it'll crash the game, likely due to a detection failing.

Reply:
I find solving that difficult. And I find that if you agree with the input filtering work, then the original way is the simplest way (and it also solves the paste problem too).

By the way, I still don't quite understand what's the problem of defocus.

And if you still prefer the enter way, we can actually add a tooltip saying that the color will only be applied when you <kbd>enter</kbd> to ignore the focus-exit problem (although it's a little irresponsible).

Reply:
So as I understand it, the focus exit problem is where the line edit doesn't lose focus when you click at a general area outside of the html color edit area unless it's a button or other line edits, etc. and as a result the color can't be applied automatically? This is fine, the player should already by intuition presses <kbd>Enter</kbd> every time they finished inputting a html color code. The automatic color apply if the html edit loses focus is just an additional safeguard. If you look at Godot's default color picker behavior, that's what it does.

I still prefer the enter way as it's the simplest implementation (doesn't require manual input filtering and else), but if you want to go otherwise that's fine.